File: typescript/packages/jumpstarter-client/src/DriverReport.ts
/**
 * Represents an exporter's device tree as returned by `ExporterService.GetReport`.
 */

/** A single driver instance in the device tree. */
export interface DriverInstance {
  /** Unique ID of this driver instance within the exporter. */
  readonly uuid: string;
  /** Parent driver UUID, or undefined for root-level drivers. */
  readonly parentUuid?: string;
  /** Labels including `jumpstarter.dev/name` and `jumpstarter.dev/client`. */
  readonly labels: ReadonlyMap<string, string>;
  /** Human-readable description, or undefined. */
  readonly description?: string;
  /** Map of method name to help text. */
  readonly methodsDescription: ReadonlyMap<string, string>;
  /** Fully-qualified gRPC service names supported by this driver. */
  readonly nativeServices: readonly string[];
}

/** The full device tree report from an exporter. */
export class DriverReport {
  /** The exporter's root UUID. */
  readonly uuid: string;
  /** The exporter's labels. */
  readonly labels: ReadonlyMap<string, string>;
  /** All driver instances in the device tree. */
  readonly instances: readonly DriverInstance[];

  constructor(response: {
    uuid: string;
    labelsMap: Array<[string, string]>;
    reportsList: Array<{
      uuid: string;
      parentUuid?: string;
      labelsMap: Array<[string, string]>;
      description?: string;
      methodsDescriptionMap: Array<[string, string]>;
      nativeServicesList: string[];
    }>;
  }) {
    this.uuid = response.uuid;
    this.labels = new Map(response.labelsMap);
    this.instances = response.reportsList.map((r) => ({
      uuid: r.uuid,
      parentUuid: r.parentUuid || undefined,
      labels: new Map(r.labelsMap),
      description: r.description || undefined,
      methodsDescription: new Map(r.methodsDescriptionMap),
      nativeServices: r.nativeServicesList,
    }));
  }

  /**
   * Find a driver instance by its `jumpstarter.dev/name` label.
   *
   * @param name - The driver name
   * @returns The matching instance, or undefined if not found
   */
  findByName(name: string): DriverInstance | undefined {
    return this.instances.find(
      (inst) => inst.labels.get("jumpstarter.dev/name") === name,
    );
  }

  /**
   * Find a driver instance by its `jumpstarter.dev/name` label, throwing
   * if not found.
   *
   * @param name - The driver name
   * @returns The matching instance
   * @throws Error if no driver with this name exists
   */
  requireByName(name: string): DriverInstance {
    const inst = this.findByName(name);
    if (!inst) {
      const available = this.instances
        .map((i) => i.labels.get("jumpstarter.dev/name") ?? i.uuid)
        .join(", ");
      throw new Error(
        `No driver found with name: ${name}. Available: [${available}]`,
      );
    }
    return inst;
  }
}

File: java/jumpstarter-client/src/main/java/dev/jumpstarter/client/DriverReport.java
package dev.jumpstarter.client;

import jumpstarter.v1.Jumpstarter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * Represents an exporter's device tree as returned by {@code ExporterService.GetReport}.
 */
public final class DriverReport {

    private final String uuid;
    private final Map<String, String> labels;
    private final List<DriverInstance> instances;

    DriverReport(@NotNull Jumpstarter.GetReportResponse response) {
        this.uuid = response.getUuid();
        this.labels = Collections.unmodifiableMap(response.getLabelsMap());
        List<DriverInstance> list = new ArrayList<>();
        for (Jumpstarter.DriverInstanceReport r : response.getReportsList()) {
            list.add(new DriverInstance(r));
        }
        this.instances = Collections.unmodifiableList(list);
    }

    /** The exporter's root UUID. */
    @NotNull
    public String getUuid() {
        return uuid;
    }

    /** The exporter's labels. */
    @NotNull
    public Map<String, String> getLabels() {
        return labels;
    }

    /** All driver instances in the device tree. */
    @NotNull
    public List<DriverInstance> getInstances() {
        return instances;
    }

    /**
     * Find a driver instance by its {@code jumpstarter.dev/name} label.
     *
     * @param name the driver name
     * @return the matching instance, or null if not found
     */
    @Nullable
    public DriverInstance findByName(@NotNull String name) {
        for (DriverInstance instance : instances) {
            String n = instance.getLabels().get("jumpstarter.dev/name");
            if (name.equals(n)) {
                return instance;
            }
        }
        return null;
    }

    /**
     * Represents a single driver instance in the device tree.
     */
    public static final class DriverInstance {
        private final String uuid;
        private final String parentUuid;
        private final Map<String, String> labels;
        private final String description;
        private final Map<String, String> methodsDescription;

        DriverInstance(@NotNull Jumpstarter.DriverInstanceReport report) {
            this.uuid = report.getUuid();
            this.parentUuid = report.hasParentUuid() ? report.getParentUuid() : null;
            this.labels = Collections.unmodifiableMap(report.getLabelsMap());
            this.description = report.hasDescription() ? report.getDescription() : null;
            this.methodsDescription = Collections.unmodifiableMap(report.getMethodsDescriptionMap());
        }

        /** Unique ID of this driver instance within the exporter. */
        @NotNull
        public String getUuid() {
            return uuid;
        }

        /** Parent driver UUID, or null for root-level drivers. */
        @Nullable
        public String getParentUuid() {
            return parentUuid;
        }

        /** Labels including {@code jumpstarter.dev/name} and {@code jumpstarter.dev/client}. */
        @NotNull
        public Map<String, String> getLabels() {
            return labels;
        }

        /** Human-readable description, or null. */
        @Nullable
        public String getDescription() {
            return description;
        }

        /** Map of method name to help text. */
        @NotNull
        public Map<String, String> getMethodsDescription() {
            return methodsDescription;
        }

        @Override
        public String toString() {
            String name = labels.getOrDefault("jumpstarter.dev/name", uuid);
            return "DriverInstance{name=" + name + ", uuid=" + uuid + "}";
        }
    }
}
